pkg/cluster/bootstrap.go

```diff
@@ -33,22 +38,39 @@ func (c *Cluster) Bootstrap(ctx context.Context) error {
 	return nil
 }
 
-func (c *Cluster) shouldBootstrapLoad() (bool, error) {
+// shouldBootstrapLoad returns true if we need to load ControlRuntimeBootstrap data again.
+// This is controlled by a stamp file on disk that records successful bootstrap using a hash of the join token.
+func (c *Cluster) shouldBootstrapLoad(ctx context.Context) (bool, error) {
+	// If using a managed database, see if we need to bootstrap
 	if c.managedDB != nil {
 		c.runtime.HTTPBootstrap = true
+		// No URL to join, so don't need to bootstrap
 		if c.config.JoinURL == "" {
 			return false, nil
 		}
 
-		token, err := clientaccess.NormalizeAndValidateTokenForUser(c.config.JoinURL, c.config.Token, "server")
+		// Validate the join token
+		info, err := clientaccess.ParseAndValidateTokenForUser(c.config.JoinURL, c.config.Token, "server")
 		if err != nil {
+			// If we got a Service Unavailable error from the join URL, and the managed database is already
+			// initialized, bypass additional bootstrap checks.
+			// Normally this would be fatal, but in the case of quorum loss the other server cannot serve the
+			// CA bundle until the datastore is up, and the datastore can't come up until we come up to restore quorum.
+			// Break this deadlock by ignoring the failure if the managed database has already been initialized.
+			// If the user has changed the JoinURL to target a different cluster and we SHOULD re-bootstrap,
+			// etcd will fail to join the new cluster due to invalid certificates, and startup will fail until
+			// another node can come online to restore quorum, or the cluster is reset by the administrator.
+			if errors.Is(err, clientaccess.ErrServiceUnavailable) {
+				if isInitialized, err2 := c.managedDB.IsInitialized(ctx, c.config); err2 != nil {
```

Comment:
can we also here just check the bootstrap file instead of making a call to the etcd? also was wondering if the etcd will respond when it is in a state of quorum loss here

Reply:
IsInitialized doesn't actually call into etcd, it just checks to see if the WAL dir exists:
https://github.com/rancher/k3s/blob/master/pkg/etcd/etcd.go#L124

Reply:
Should we just move the bootstrapStamp() check earlier?  Wouldn't that work around this?

Reply:
@ibuildthecloud  I tried to test this idea. And in my case, I have server dns name map to all IPs address of master. I mean `--server <servers-dns>:6443` instead of `--server <ip address of the first server>:6443`  I then endup in having 3 clusters with single master instead of having a cluster with 3 master. Not sure if I have things up in correct way.

Reply:
@ibuildthecloud I think that it's a good idea to always check the token BEFORE the bootstrap stamp to ensure that it's still valid. This catches and gives meaningful errors for some odd corner cases like users bringing up old nodes that think they're still connected to clusters that have since been rebuilt. I only want to skip the token checks IFF the master server is giving 503s, and we are an initialized etcd member.

I think that just moving the check up would fix the immediate issue, but corner cases like the one mentioned above will fail with odd errors from etcd, instead of a nicer message about the cluster token.

Reply:
Okay, I think I get what your saying. My only comments it that keying the behavior off of 503 (which is a pretty generic error) seems arbitrary.  If you bootstrap twice by accident is there a downside to it?

Reply:
503 is what we send in our default handler, which only ends up being used while the managed database is starting up. We could send something more specific, or send a specific payload that was also checked for?

I think re-bootstrapping should be safe. It's easy enough to do by deleting the file, or switching between using just the password and the the full string that includes the hash and username - since the bootstrap stamp is based on the string provided by the user.